API `byIDs` added to `ArrayList`

dhensby opened the pull request:
Added `byIDs` for consistent API with `DataList`

dhensby commented:
@silverstripe/core-team to be honest, I'm thinking that this function `byIDs` on DataList/ArrayList should just be deprecated.

It's a function that is effectively a hard coded filter which doesn't add any real benefit over doing `$this->filter('ID', $ids);`

Having it as part of the code base increases code maintenance and doesn't really provide any special flexibility.

tractorcow commented:
It's probably ok, since filtering by ID (in `DataList`) negates the necessity to do a classwalk to find the table for the filtered field. It doesn't really give the same benefit in ArrayList as it does there though.

halkyon commented on tests/model/ArrayListTest.php:
```diff
@@ -746,7 +746,8 @@ public function testCanFilterByEmpty() {
 
-	public function testByID() {
+	public function testByID()
+	{
```
*super nitpicky mode* brace shouldn't be on a newline ;-)

tractorcow replied:
I had `setenv nitpickymode 0`, but yeah I saw that. :)

Welcome back @halkyon

dhensby replied:
PHPStorm! Rage! Can I get away with saying.... PSR?

stevie-mayhew replied:
Wasn't it you who added the .editorconfig files @dhensby ? :D

dhensby replied:
Unfortunately you can't dictate code style like this, it's more white space based. :(

tractorcow commented:
The issue with travis is caused by travis-support generating the incorrect `version` tags for framework on the 3 branch (it should be 3.x-dev not dev-3). I'll look into this later. :D

dhensby commented:
Ok, great. I've fixed that naughty curly brace going to the wrong line and I've also squashed the tests

dhensby commented:
Ok, so I noticed that the `SS_Filterable` interface has diverged a bit from what's implemented on `DataList` and `ArrayList` so I've updated the interface and implemented the method(s) as needed on `ArrayList`.

I've also opened issue #3847 so we can have some clarity on what should be returned by the `map` method.

As I've now edited the interface, I think that either this needs to go against master OR I need to pull out the interface changes and add them to master.

dhensby commented:
As a compromise we could put all these changes into 3 (except the interface changes) and then put the interface changes into 4.

Therefore we get a consistent API which is BC for 3.2 and then enforce the interface properly (non-BC) in 4?

dhensby commented:
bump @silverstripe/core-team

camspiers commented:
As mentioned on #4557 I think there is a mistake in ArrayList::filterAny, and I think tests should be added for it.

tractorcow commented on model/ArrayList.php:
```diff
@@ -477,6 +477,72 @@ public function canFilterBy($by) {
+
+	}
+
+	protected function normaliseFilterArgs() {
```
phpdoc please.

tractorcow commented on model/ArrayList.php:
```diff
@@ -477,6 +477,72 @@ public function canFilterBy($by) {
+	 * @return DataList
+	 */
+	public function filterAny() {
+		$keepUs = call_user_func_array(array($this, 'normaliseFilterArgs'), func_get_args());
```
Why not just make normaliseFilterArgs take a single $args parameter? code inspection tools hate code linked with call_user_func_array(), and it's a bit ugly.

dhensby replied:
Because the number of arguments changes the behaviour of these functions. the point is to turn that unknown number of args into a normalised array

tractorcow replied:
Yeah, and that number will just be the size of the array that's passed in.

```php
$keepUs = $this->normaliseFilterArgs(func_get_args());
```

and
```php
function normaliseFilterArgs($args) {}
```

dhensby replied:
hmmm. *personally* I don't like that, it's yet another method api convention that's different...

I suppose my personal preference here shouldn't make a big difference; if it makes a difference to code inspectio... i suppose it's worth it.

tractorcow replied:
ok, fine. :D

tractorcow commented on model/ArrayList.php:
```diff
@@ -477,6 +477,72 @@ public function canFilterBy($by) {
+
+		foreach ($this->items as $column => $value) {
+			$keepItem = false;
+			if ((is_array($value) && in_array($this->extractValue($item, $column), $value))
```
```php
$extractedValue = $this->extractValue($item, $column);
$matches = is_array($value) ? in_array($extractedValue, $value) : $extractedValue == $value;
if($matches) {...}
```

Cut down on duplicate code. ;)

tractorcow replied:
you may as well put it in a method for checking matches, then having all the filter methods call it. It seems to be duplicated logic in each? (exclude, filter, filterAny, etc).

dhensby replied:
The matching changes slightly if you want to match any or all, though.

It would be better to fix that, but at the moment I just want to add these APIs in rather than refactor large amounts of `ArrayList`

tractorcow commented on model/UnsavedRelationList.php:
```diff
@@ -299,7 +299,7 @@ public function avg() {
 	}
 
-	public function byIDs() {
+	public function byIDs(array $ids) {
```
Is this going to break semver for subclasses?

dhensby replied:
If someone's subclassed `UnsavedRelationList` and overloaded `byIDs` and NOT accepted an ID parameter... well... then I don't know what kind of world we are living in.

I suppose I could remove the type hinting if we thought there was a real risk of regression here; personally, I don't

tractorcow replied:
Yeah, we probably should punish poor coding. :)

dhensby commented:
feedback dealt with

camspiers commented on model/ArrayList.php:
```diff
@@ -477,6 +477,80 @@ public function canFilterBy($by) {
+		$itemsToKeep = array();
+
+		foreach ($this->items as $item) {
+			$keepItem = false;
```
looks like you can remove this now.

dhensby replied:
removed, thanks :)

tractorcow commented:
Dealings-with with fed-back

sminnee commented:
Hey @dhensby @tractorcow where are we at with this?

tractorcow commented:
We are at merged.

dhensby commented:
yay, thanks

camspiers commented:
Nice work @dhensby!
